libs/pq_proto/src/framed.rs

```diff
@@ -0,0 +1,253 @@
+}
+
+impl<S: AsyncRead + AsyncWrite + Unpin> Framed<S> {
+    /// Split into owned read and write parts. Beware of potential issues with
+    /// using halves in different tasks on TLS stream:
+    /// https://github.com/tokio-rs/tls/issues/40
+    pub fn split(self) -> (FramedReader<ReadHalf<S>>, FramedWriter<WriteHalf<S>>) {
+        let (read_half, write_half) = tokio::io::split(self.stream);
+        let reader = FramedReader {
+            stream: read_half,
+            read_buf: self.read_buf,
+            startup_read: self.startup_read,
+        };
+        let writer = FramedWriter {
+            stream: write_half,
+            write_buf: self.write_buf,
+        };
+        (reader, writer)
+    }
+
+    // Join read and write parts back.
+    pub fn unsplit(reader: FramedReader<ReadHalf<S>>, writer: FramedWriter<WriteHalf<S>>) -> Self {
+        Self {
+            stream: reader.stream.unsplit(writer.stream),
+            read_buf: reader.read_buf,
+            write_buf: writer.write_buf,
+            startup_read: reader.startup_read,
+        }
+    }
+}
+
+/// Read-only version of `Framed`.
+pub struct FramedReader<S> {
+    stream: S,
+    read_buf: BytesMut,
+    // Have we already decoded startup message? All further should start with
+    // message type byte then.
+    startup_read: bool,
+}
+
+impl<S: AsyncRead + Unpin> FramedReader<S> {
+    pub async fn read_message(&mut self) -> Result<Option<FeMessage>, ConnectionError> {
+        read_message(&mut self.stream, &mut self.read_buf, &mut self.startup_read).await
+    }
+}
+
+/// Write-only version of `Framed`.
+pub struct FramedWriter<S> {
+    stream: S,
+    write_buf: BytesMut,
+}
+
+impl<S> FramedWriter<S> {
+    /// Get a mut reference to the underlying stream.
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.stream
+    }
+}
+
+impl<S: AsyncWrite + Unpin> FramedWriter<S> {
+    pub fn write_message(&mut self, msg: &BeMessage<'_>) -> Result<(), ProtocolError> {
+        write_message(&mut self.write_buf, msg)
+    }
+
+    pub async fn flush(&mut self) -> Result<(), io::Error> {
+        flush(&mut self.stream, &mut self.write_buf).await
+    }
+}
+
+/// Read next message from the stream. Returns Ok(None), if EOF happened and we
+/// don't have remaining data in the buffer. This function is cancellation safe:
+/// you can drop future which is not yet complete and finalize reading message
+/// with the next call.
+async fn read_message<S: AsyncRead + Unpin>(
+    stream: &mut S,
+    read_buf: &mut BytesMut,
+    startup_read: &mut bool,
+) -> Result<Option<FeMessage>, ConnectionError> {
+    loop {
+        if let Some(msg) = decode(read_buf, startup_read)? {
+            return Ok(Some(msg));
+        }
+        // If we can't build a frame yet, try to read more data and try again.
+        // Make sure we've got room for at least one byte to read to ensure
+        // that we don't get a spurious 0 that looks like EOF.
+        read_buf.reserve(1);
```

Comment:
This piece was actually directly copied from
https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#220

AFAIU partially this `reserve(1)` is more like guard against badly behaving codec; often codec requests chunk big enough to get the whole packet and this is no op. Sometimes it might not know it (not even first bytes of packet were received), and buffer can be indeed full. However, even in this case this doesn't mean inefficient realloc of +1 bytes size; `BytesMut` will first try to shift data left, occuping already released space in the buffer, and most likely will succeed in that if we Drop old messages. That's my general understanding, I haven't studied `bytes` in detail though; but I'm also curious about that.

> Also, read_buf has a strange [doc](https://docs.rs/tokio/latest/tokio/io/trait.AsyncReadExt.html#return-1)

Some streams, in theory, e.g. something in memory can produce EOF and then again data if something changes. This is definitely not the case for `TcpStream`, but this is generic func.